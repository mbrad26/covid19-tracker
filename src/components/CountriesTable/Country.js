import React from 'react';
import { Image } from 'react-bootstrap';

const Country = ({
  country,
  setActive,
  setInactive,
  dataSuccess,
  zoomOnCountry,
  updateEndPoint,
  historicalCountryDataLoading,
  }) => {

  const handleClick = () => {
    dataSuccess(country);
    updateEndPoint(country.country);
    historicalCountryDataLoading();
    zoomOnCountry(country);
    setInactive();
    setTimeout(() => {
      setActive(country);
    }, 2000);
  };

  return (
    <>
      <tbody>
        <tr>
          <td><Image
                src={country.countryInfo.flag}
                loading='lazy'
                alt={`${country.country} flag`}
                id='flag'
              />
          </td>
          <td>
            <button
              onClick={handleClick}
            >
              {country.country}
            </button>
          </td>
          <td>{country.cases.toLocaleString()}</td>
          <td>{country.deaths.toLocaleString()}</td>
          <td>{country.recovered.toLocaleString()}</td>
        </tr>
      </tbody>
    </>
  );
};

export default Country;
