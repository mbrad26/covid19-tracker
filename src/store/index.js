import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { createLogger } from 'redux-logger';
import rootReducer from '../reducers';
import rootSaga from '../sagas';

const logger = createLogger();
const saga = createSagaMiddleware();

const store = createStore(
  rootReducer,
  applyMiddleware(saga, logger)
);

// console.log('STORE: ', store.getState());

saga.run(rootSaga);

export default store;
