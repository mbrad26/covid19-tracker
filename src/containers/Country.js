import { connect } from 'react-redux';

import Country from '../components/CountriesTable/Country';
import { doDataSuccess } from '../actions/globalData';
import { doZoomOnCountry } from '../actions/mapboxData';
import { doUpdateEndPoint, doHistoricalCountryDataLoading } from '../actions/historicalData';
import { doSetActive, doSetInactive } from '../actions/countryData';

const mapDispatchToProps = dispatch => ({
  setInactive: () => dispatch(doSetInactive()),
  dataSuccess: data => dispatch(doDataSuccess(data)),
  setActive: country => dispatch(doSetActive(country)),
  zoomOnCountry: country => dispatch(doZoomOnCountry(country)),
  updateEndPoint: endPoint => dispatch(doUpdateEndPoint(endPoint)),
  historicalCountryDataLoading: () => dispatch(doHistoricalCountryDataLoading()),
});

export default connect(null, mapDispatchToProps)(Country);
