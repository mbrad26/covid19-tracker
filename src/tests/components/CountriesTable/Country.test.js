import React from 'react';
import { Provider } from 'react-redux';
import { render, screen, fireEvent } from '@testing-library/react';

import store from '../../../store';
import { country } from '../../fixtures';
import Country from '../../../components/CountriesTable/Country';

describe('Country', () => {
  const countryProps = {
    country: country,
    setActive: jest.fn(),
    setInactive: jest.fn(),
    dataSuccess: jest.fn(),
    zoomOnCountry: jest.fn(),
    updateEndPoint: jest.fn(),
    historicalCountryDataLoading: jest.fn(),
  }

  const table = document.createElement('table');

  beforeEach(() => {
    jest.useFakeTimers();
    render(
      <Provider store={store}>
        <Country {...countryProps}/>
      </Provider>, {
        container: document.body.appendChild(table),
      }
    );
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });


  it('renders snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <Country {...countryProps} />
      </Provider>, {
        container: document.body.appendChild(table),
      }
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders all the props', () => {
    expect(screen.getByText('284')).toBeInTheDocument();
    expect(screen.getByText('9,513')).toBeInTheDocument();
    expect(screen.getByText('5,214')).toBeInTheDocument();
  });

  it('renders an image', () => {
    expect(screen.getByRole('img')).toHaveAttribute('alt', 'Albania flag');
  });

  it('renders a button', () => {
    expect(screen.getByRole('button', { name: 'Albania'})).toBeInTheDocument();
  });

  it('clicking the button calls the callback handler', () => {
    fireEvent.click(screen.getByRole('button', { name: 'Albania' }));

    expect(countryProps.dataSuccess).toHaveBeenCalledTimes(1);
    expect(countryProps.dataSuccess).toHaveBeenCalledWith(country);
    expect(countryProps.updateEndPoint).toHaveBeenCalledTimes(1);
    expect(countryProps.updateEndPoint).toHaveBeenCalledWith(country.country);
    expect(countryProps.historicalCountryDataLoading).toHaveBeenCalledTimes(1);
    expect(countryProps.zoomOnCountry).toHaveBeenCalledTimes(1);
    expect(countryProps.zoomOnCountry).toHaveBeenCalledWith(country);
    expect(countryProps.setInactive).toHaveBeenCalledTimes(1);
  });

  it('clicking the button setActive after 2 seconds', () => {
    fireEvent.click(screen.getByRole('button', { name: 'Albania' }));

    expect(countryProps.setActive).toHaveBeenCalledTimes(1);
    expect(countryProps.setActive).toHaveBeenCalledWith(country);
  });

});
