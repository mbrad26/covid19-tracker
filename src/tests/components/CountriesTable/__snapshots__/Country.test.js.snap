// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country renders snapshot 1`] = `
<tbody>
  <tr>
    <td>
      <img
        alt="Albania flag"
        class=""
        id="flag"
        loading="lazy"
        src="https://disease.sh/assets/img/flags/al.png"
      />
    </td>
    <td>
      <button>
        Albania
      </button>
    </td>
    <td>
      9,513
    </td>
    <td>
      284
    </td>
    <td>
      5,214
    </td>
  </tr>
</tbody>
`;
